fuzz/src/invoice_request_deser.rs

```diff
@@ -0,0 +1,112 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+use bitcoin::secp256k1::{KeyPair, Parity, PublicKey, Secp256k1, SecretKey, self};
+use crate::utils::test_logger;
+use core::convert::{Infallible, TryFrom};
+use lightning::chain::keysinterface::EntropySource;
+use lightning::ln::PaymentHash;
+use lightning::ln::features::BlindedHopFeatures;
+use lightning::offers::invoice::{BlindedPayInfo, UnsignedInvoice};
+use lightning::offers::invoice_request::InvoiceRequest;
+use lightning::offers::parse::SemanticError;
+use lightning::onion_message::BlindedPath;
+use lightning::util::ser::Writeable;
+
+#[inline]
+pub fn do_test<Out: test_logger::Output>(data: &[u8], _out: Out) {
+	if let Ok(invoice_request) = InvoiceRequest::try_from(data.to_vec()) {
+		let mut bytes = Vec::with_capacity(data.len());
+		invoice_request.write(&mut bytes).unwrap();
+		assert_eq!(data, bytes);
+
+		let secp_ctx = Secp256k1::new();
+		let keys = KeyPair::from_secret_key(&secp_ctx, &SecretKey::from_slice(&[42; 32]).unwrap());
+		let mut buffer = Vec::new();
+
+		if let Ok(unsigned_invoice) = build_response(&invoice_request, &secp_ctx) {
+			let signing_pubkey = unsigned_invoice.signing_pubkey();
+			let (x_only_pubkey, _) = keys.x_only_public_key();
+			let odd_pubkey = x_only_pubkey.public_key(Parity::Odd);
+			let even_pubkey = x_only_pubkey.public_key(Parity::Even);
+			if signing_pubkey == odd_pubkey || signing_pubkey == even_pubkey {
+				unsigned_invoice
+					.sign::<_, Infallible>(
+						|digest| Ok(secp_ctx.sign_schnorr_no_aux_rand(digest, &keys))
+					)
+					.unwrap()
+					.write(&mut buffer)
+					.unwrap();
```

Comment:
I'm wondering why is making sure it's possible to sign (here and in the others) important to test with fuzzing? I figure signing wouldn't really change much between invoices built from different requests?

Also on this section, could someone elaborate on what this means?

> The fuzzer should have no problem making 32 bytes of the input match something that's hard-coded, as long as the hard-coded thing is a "nice" pattern.

Reply:
> I'm wondering why is making sure it's possible to sign (here and in the others) important to test with fuzzing? I figure signing wouldn't really change much between invoices built from different requests?

It's probably not super critical, but it will exercise message-specific serialization code and use the result to form a message digest to sign. The latter uses more generic `TlvStream` code for iterating through the bytes, which expects well-formed input.

> 
> Also on this section, could someone elaborate on what this means?
> 
> > The fuzzer should have no problem making 32 bytes of the input match something that's hard-coded, as long as the hard-coded thing is a "nice" pattern.

I believe this pertains to how the fuzzer explores different code branches, but @TheBlueMatt can provide more thorough explanation.

Reply:
> It's probably not super critical, but it will exercise message-specific serialization code and use the result to form a message digest to sign. The latter uses more generic `TlvStream` code for iterating through the bytes, which expects well-formed input.

Ah gotcha

> but its not actually random, instead its a defined set of mutations/"nice patterns" (all 0s, all 1, etc, etc) made to the existing set of tests, focusing on tests that had good code coverage.

Oh woah cool I'll have to look more into this